OIDC - add Azure processing information. (#5449)
mendix/docs: content/en/docs/appstore/modules/oidc.md

### 5.1 OIDC Provider Configuration

#### 5.1.1 General OIDC Providers

1. In your IdP, provision a new OpenID client application. You will receive a ClientID and Client Secret.
2. You will also need the OIDC configuration endpoint (for example: [https://accounts.google.com/.well-known/openid-configuration](https://accounts.google.com/.well-known/openid-configuration))
3. Register the following callback URLs:
    * `https://<your-app-url>/oauth/v2/callback`
    * `makeitnative://<your-app-url>/oauth/callback`

#### 5.1.2 Azure AD Provider Configuration for APIs{#azure-portal}

> This section gives some guidance for doing the necessary configurations at your Azure AD provider to obtain access tokens for your APIs.

Reviewer: "to obtain access tokens containing the right authorisation claims to secure your APIs".

You always get an AT, but it';s about making sure it's good one ;-)

Author: Done: https://github.com/mendix/docs/pull/5449/commits/69d100a4b2f0a3e79168fe35af4ef5aae5ecbf32

> If you don’t set the access token up correctly, you will get access tokens that contain both `nonce` and `aud` claims. These are intended to be presented to the Microsoft Graph API and cannot be validated by your API.

Reviewer: "If you don't .. correctly, AzureAD will generate an access token with a value for the 'aud' claim that indicates the token is intended to be presented to the Microsoft Graph API and cannot be validated by your API."

I don't think the nonce has anything to do with this although it may be releated. Let's leave nonce out. The message to the reader doesn't chnage.

Author: Done: https://github.com/mendix/docs/pull/5449/commits/69d100a4b2f0a3e79168fe35af4ef5aae5ecbf32

To get the Microsoft Identity Platform to issue access tokens you can pass to your API, you need to set up a custom scope in the App Registration’s **Expose an API** tab, and request that scope when you acquire the tokens. TO do this:

1. Open the **Expose an API** tab in the **App Registration** page of the Azure Portal.
1. In the **Expose an API** tab, set up a custom scope.
    The scope will be prefixed with your `Application ID URI`.
1. In the **API permissions** tab, assign the created scope to the application.

By adding a custom claim to the App Registration’s Expose an API tab and requesting that scope when we acquire tokens, the Microsoft Identity Platform will now generate access tokens that we can present to our API

> 

Reviewer: Sharisma, this seems to be contradicting the other info you've added yesterday:

'Claim used from Access token to parse the roles:

1. SAM -  sws.samauth.role.name
2. PIB - scope
3. Azure - roles'

I'm confused: does the AzureAD microflow evaluate the 'scope' or the 'roles' claim' or both?

Reviewer 2: @JaapF 
AzureAD microflow evaluate 'roles' claim to parse roles.

We need to set 'custom scope' to get access token which needs to be validated by JWKS URI

Author: I've added some information - does this clarify things @JaapF ?
https://github.com/mendix/docs/pull/5449/commits/c46c86ac64939e8191be5778c1468dfa51a4add0

Reviewer: Yes it clarifies things, although we may consider enhancing/changing the AzureAD microflow.
But I think it accurately describes what we have today.
Let's be pragmatic and publish.

### 5.1 OIDC Provider Configuration

#### 5.1.1 General OIDC Providers

1. In your IdP, provision a new OpenID client application. You will receive a ClientID and Client Secret.
2. You will also need the OIDC configuration endpoint (for example: [https://accounts.google.com/.well-known/openid-configuration](https://accounts.google.com/.well-known/openid-configuration))
3. Register the following callback URLs:
    * `https://<your-app-url>/oauth/v2/callback`
    * `makeitnative://<your-app-url>/oauth/callback`

#### 5.1.2 Azure AD Provider Configuration for APIs{#azure-portal}

This section gives some guidance for doing the necessary configurations at your Azure AD provider to obtain access tokens containing the right authorization claims to secure your APIs.

> If you don’t set the access token up correctly, you will get access tokens containing `aud` claims. These are intended to be presented to the Microsoft Graph API and cannot be validated by your API.

Reviewer: explanation to mark:
The value of the aud(ience) claims indicate the audience for the token.
By default the audience (apparantly) is the microsoft graph-API.
For a different audience (e.g. mendix API), configuration is necessary.

so the issue is not getting the aud claim, but getting aud claim with wrong value.
as a consequence the token cannot be introspected/validated by the wrong audience.

Author: Updated to reflect Jaap's explanation.
https://github.com/mendix/docs/pull/5449/commits/fa9f7301aa7b5121af5ee931516ac33048b150e4

3. Register the following callback URLs:
    * `https://<your-app-url>/oauth/v2/callback`
    * `makeitnative://<your-app-url>/oauth/callback`

#### 5.1.2 Azure AD Provider Configuration for APIs{#azure-portal}

This section gives some guidance for doing the necessary configurations at your Azure AD provider to obtain access tokens containing the right authorization claims to secure your APIs.

If you don’t set the access token up correctly, you will get access tokens containing `aud` claims. These are intended to be presented to the Microsoft Graph API and cannot be validated by your API.

To get the Microsoft Identity Platform to issue access tokens you can pass to your API, you need to set up a custom scope in the App Registration’s **Expose an API** tab, and request that scope when you acquire the tokens. TO do this:

1. Open the **Expose an API** tab in the **App Registration** page of the Azure Portal.
1. In the **Expose an API** tab, set up a custom scope.
    The scope will be prefixed with your `Application ID URI`.
1. In the **API permissions** tab, assign the created scope to the application.

> 1. In the **App roles** tab, add the user roles you want to authorize using either the user role name, or the user role UUID.

Reviewer 2: @MarkvanMents Shall we add the below info?
Configured user roles in Azure will be included in 'roles' claim of access token of Azure

Author: Updated. https://github.com/mendix/docs/pull/5449/commits/fa9f7301aa7b5121af5ee931516ac33048b150e4

> By adding a custom claim to the App Registration’s Expose an API tab and requesting that scope when we acquire tokens, the Microsoft Identity Platform will now generate access tokens that we can present to our API

Reviewer 2: @MarkvanMents Can we add some more specific information like "we need to configure scope in Azure portal to get access tokens validated by JWKS URI"

Author: Updated. https://github.com/mendix/docs/pull/5449/commits/fa9f7301aa7b5121af5ee931516ac33048b150e4

This section gives some guidance for doing the necessary configurations at your Azure AD provider to obtain access tokens containing the right authorization claims to secure your APIs.

If you don’t set the access token up correctly, you will get access tokens containing default `aud` (audience) claims. The default audience is the Microsoft Graph API and so these access tokens cannot be validated by your API.

To get the Microsoft Identity Platform to issue access tokens you can pass to your API, you need to set up a custom scope in the App Registration’s **Expose an API** tab, and request that scope when you acquire the tokens. TO do this:

1. Open the **Expose an API** tab in the **App Registration** page of the Azure Portal.
1. In the **Expose an API** tab, set up a custom scope.
    The scope will be prefixed with your `Application ID URI`.
1. In the **API permissions** tab, assign the created scope to the application.
1. In the **App roles** tab, add the user roles you want to authorize using either the user role name, or the user role UUID. This adds the configured user roles to the roles claim in the access token. 

> By adding a custom claim to the App Registration’s Expose an API tab and requesting that scope when we acquire tokens, the Microsoft Identity Platform will now generate access tokens that can be validated at the `/jwks` URI.

Reviewer: "validated using info retrieved from the /jwks URI." 
the endpoint itself doesn't validate the token.

Author: Does this suggestion make sense? Or can you suggest a better turn of phrase?
```suggestion
By adding a custom claim to the App Registration’s Expose an API tab and requesting that scope when we acquire tokens, the Microsoft Identity Platform will now generate access tokens that can be validated using the `/jwks` URI.
```

The OIDC SSO module supports multiple IdPs. Since each provider can provide user data in a different format, you may want to use multiple provisioning flows. See the microflow `UserProvisioning_Sample` for an example and details on how to do this.

## 7 API Authentication {#api-authentication}

> You can create your own APIs within your Mendix app and secure the end point over OICD using a custom authentication microflow. To do this:

Reviewer: typo OIDC not OICD

Reviewer: One could say that OIDC is an enhanced version of OAuth. Simple explanation:
OAuth is delegating Authorization decision
OIDC is delegating Authentication.
The scope/AT-parsing stuff is about delegating authorization, so it would be more correct to say 'OAuth'.
We could also be pragmatic and say "OAuth/OIDC" not to bother the reader with this nuance.

Author: > We could also be pragmatic and say "OAuth/OIDC" not to bother the reader with this nuance.

We do say this in section 8.2 (Access Token Parsing) - "With the OAuth/OIDC protocol, access tokens can be opaque or can be a JSON Web Token (JWT)."

We also say in the introduction "OIDC is an extension of OAuth2 that propagates the end-user's identity to your application." Is there anywhere else that we need to make a distinction?

The microflow returns an object of type `System.HttpResponse`. This could indicate an error.

### 8.3 Access Token Parsing{#access-token-parsing}

With the OAuth/OIDC protocol, access tokens can be opaque or can be a JSON Web Token (JWT).
If you are just delegating authentication for your app to the IdP you will not need to know the contents of the access token.

If you want to use the information in an access token which is a JWT, you need to parse the access token in a microflow. For example, you may want to assign user roles in your app based on the contents of the access token JWT.

* The OIDC module provides you with default microflows for parsing access tokens from the following IdPs:
    * Siemens SAM – in this case the `sws.samauth.role.name` claim is interpreted
    * AzureAD – in this case the `roles` claim is interpreted
    * Private IAM Broker (PIB) – in this case the `scope` claim is interpreted

> * If you are using another IdP or want to use a different claim, you can create a custom microflow to parse the access token.

Reviewer: Brief explanation about PIB.

"Mendix will be releasing a so-called Private IAM Broker (PIB), which will allow external non-Mendix clients to request Access Tokens so they can consume APIs exposed by your Mendix backend app without any dependency on other IAM infratructure."

The PIB will support more functionality later on, but this describes the first release (MVP).

Author: Done: https://github.com/mendix/docs/pull/5449/commits/69d100a4b2f0a3e79168fe35af4ef5aae5ecbf32

1. Select the appropriate microflow to parse the access token as described in the relevant section below. If you have added a new microflow, you will need to refresh the module containing your microflow as described in [Installing Mx Model Reflection](#mxmodelreflection).

#### 8.2.1 Parsing SAM Access Tokens

{{% alert color="info" %}}
This section is only relevant if you are a Mendix partner and you want to integrate your app with the Siemens SAM IdP.
{{% /alert %}}

To parse of SAM access tokens you need to do the following when performing [OIDC Client Configuration](#client-configuration):

1. Select *Default SAM Token processing* as the **custom AccessToken processing microflow**.
    {{< figure src="/attachments/appstore/modules/oidc/enable-sam-parsing.png" >}}
1. Add the scopes `sam_account`, `samauth.role`, `samauth.tier`, and `samauth.ten` to the **Selected Scopes** in the OIDC Client Configuration.
1. Configure the user roles in your app to match the roles returned by SAM. End-users will be given the matching role when they sign into the app. If the role in the SAM token is not found in the Mendix app the end-user will be given the role `User`.
1. Save the configuration.

> 

Reviewer: @MarkvanMents 
We don't have a section on AzureAD AT parsing yet.
I think we should add that to be consistent.

@Sharishma can you help to add some details? Also because I'm confused about 'scope' versus 'roles' claim.

I'm assuming you can map values of the 'roles' claim to the userroles.

Reviewer 2: Yes @JaapF we map values of the 'roles' claim to the userroles.

@MarkvanMents I also think -a separate section for Azure needs to be added in the document as we are providing default microflow for azure as well

Author: Added. https://github.com/mendix/docs/pull/5449/commits/fa9f7301aa7b5121af5ee931516ac33048b150e4

The microflow returns an object of type `System.HttpResponse`. This could indicate an error.

### 8.2 Access Token Parsing{#access-token-parsing}

With the OAuth/OIDC protocol, access tokens can be opaque or can be a JSON Web Token (JWT).
If you are just delegating authentication for your app to the IdP you will not need to know the contents of the access token.

If you want to use the information in an access token which is a JWT, you need to parse the access token in a microflow. For example, you may want to assign user roles in your app based on the contents of the access token JWT.

* The OIDC module provides you with default microflows for parsing access tokens from the following IdPs:
    * Siemens SAM – in this case the `sws.samauth.role.name` claim is interpreted
    * AzureAD – in this case the `roles` claim is interpreted

>     * Private IAM Broker (PIB) – in this case the `scope` claim is interpreted

Reviewer 2: Examples for structures(claims) of IDPs

PIB
{
  "scope": "mx:app:userrole:0b016a49-d6da-4a2b-9b03-1a8e699d66fe mx:app:userrole:46fcedb0-2352-46fa-815d-ad2a0bcba5d8 mx:app:userrole:8b42e500-c1fa-48fe-bc5a-4f389110f045"
}

azure
"roles": [
    "53f5d6fa-6da9-4a71-b011-454ec052cce8"
  ]

SAM
"sws.samauth.role.name": [
    "53f5d6fa-6da9-4a71-b011-454ec052cce8"
  ]

Author: Examples added. https://github.com/mendix/docs/pull/5449/commits/fa9f7301aa7b5121af5ee931516ac33048b150e4

* If you are using another IdP or want to use a different claim, you can create a custom microflow to parse the access token.

To parse access tokens, you need to do the following:

1. Create a secure REST API endpoint following the instructions in [API Authentication](#api-authentication), above.
1. Run your app and sign in as an administrator, for example `Demo_administrator`.
1. Configure the client information in the OIDC Client configuration screen.
1. Check **Enable Access Token Parsing** to parse access tokens when performing [OIDC Client Configuration](#client-configuration).
1. Select the appropriate microflow to parse the access token as described in the relevant section below. If you have added a new microflow, you will need to refresh the module containing your microflow as described in [Installing Mx Model Reflection](#mxmodelreflection).

#### 8.2.1 Parsing SAM Access Tokens

{{% alert color="info" %}}
This section is only relevant if you are a Mendix partner and you want to integrate your app with the Siemens SAM IdP.
{{% /alert %}}

To parse of SAM access tokens you need to do the following when performing [OIDC Client Configuration](#client-configuration):

> 1. Select *Default SAM Token processing* as the **custom AccessToken processing microflow**.

Reviewer 2: The name of the microflows for 
1. SAM is "Default_SAM_TokenProcessing_CustomATP"
2. PIB is "Default_PIB_TokenProcessing_CustomATP"
3. Azure is "Default_Azure_TokenProcessing_CustomATP"

Reviewer 2: @MarkvanMents Here we need to update the image as well
<img width="1708" alt="image" src="https://user-images.githubusercontent.com/67798781/210220644-82a6b7ca-6089-49a9-b7be-416dfaef6308.png">

Author: Updated: https://github.com/mendix/docs/pull/5449/commits/fa9f7301aa7b5121af5ee931516ac33048b150e4

    {{< figure src="/attachments/appstore/modules/oidc/enable-sam-parsing.png" >}}
1. Add the scopes `sam_account`, `samauth.role`, `samauth.tier`, and `samauth.ten` to the **Selected Scopes** in the OIDC Client Configuration.
1. Configure the user roles in your app to match the roles returned by SAM. End-users will be given the matching role when they sign into the app. If the role in the SAM token is not found in the Mendix app the end-user will be given the role `User`.
1. Save the configuration.

#### 8.2.2 Parsing PIB Access Tokens

Mendix has released a Private IAM Broker (PIB). This allows non-Mendix clients to request access tokens which authorize them to consume APIs exposed by your Mendix backend app without a dependency on any other IAM infrastructure.

The OIDC SSO module provides a default access token parsing microflow for PIB. To use it, select the appropriate access token parsing microflow:

* For PIB, the default access token parsing microflow is `OIDC.Default_PIB_TokenProcessing_CustomATP`.

> * You can also parse the access token using the custom microflow `OIDC.ACT_Token_CustomATPRetrieveRoles` which takes an access token as input and returns a list of non-persistent objects of type `OIDC.Role`. See the next section for more information.

Author: Why do we mention using a custom microflow specifically for PIB but not for any of the others. Is there a special use case for PIB, in which case, we should perhaps mention it, or should we remove it from here as well and just have the separate section?

Reviewer 2: @MarkvanMents custom microflow"OIDC.ACT_Token_CustomATPRetrieveRoles" is general for all. Everyone can use it to customise for the token parsing

Reviewer 2: So It is better to add it as a separate section

To confirm that the authorization is working, get an access token from PIB and pass it to the API Endpoint using the authorization header. You can use Postman or any client application.

#### 8.2.3 Parsing Access Tokens Using a Custom Microflow 

> If you choose to implement your own microflow to parse an access token, the microflow name must contain `CustomATP`, for example `CustomATP_MyTokenParser`. This is how you can parse access tokens issued by IdPs such as Microsoft Azure.

Reviewer 2: @MarkvanMents ,
OIDC module contains default microflow for Azure as well
Default microflow "Default_Azure_TokenProcessing_CustomATP" is included in the Azure folder of the module for Azure

Author: Added. https://github.com/mendix/docs/pull/5449/commits/fa9f7301aa7b5121af5ee931516ac33048b150e4

* If you are using another IdP or want to use a different claim, you can create a custom microflow to parse the access token.

To parse access tokens, you need to do the following:

1. Create a secure REST API endpoint following the instructions in [API Authentication](#api-authentication), above.
1. Run your app and sign in as an administrator, for example `Demo_administrator`.
1. Configure the client information in the OIDC Client configuration screen.
1. Check **Enable Access Token Parsing** to parse access tokens when performing [OIDC Client Configuration](#client-configuration).
1. Select the appropriate microflow to parse the access token as described in the relevant section below. If you have added a new microflow, you will need to refresh the module containing your microflow as described in [Installing Mx Model Reflection](#mxmodelreflection).

#### 8.2.1 Parsing SAM Access Tokens

{{% alert color="info" %}}
This section is only relevant if you are a Mendix partner and you want to integrate your app with the Siemens SAM IdP.
{{% /alert %}}

To parse of SAM access tokens you need to do the following when performing [OIDC Client Configuration](#client-configuration):

> 1. Select *Default SAM Token processing* as the **custom AccessToken processing microflow**.

Reviewer 2: @MarkvanMents Can you please change the name of the microflow to 'OIDC.Default_SAM_TokenProcessing_CustomATP'?

Author: ```suggestion
1. Select *OIDC.Default_SAM_TokenProcessing_CustomATP* as the **custom AccessToken processing microflow**.
```

    {{< figure src="/attachments/appstore/modules/oidc/enable-sam-parsing.png" >}}
1. Add the scopes `sam_account`, `samauth.role`, `samauth.tier`, and `samauth.ten` to the **Selected Scopes** in the OIDC Client Configuration.
1. Configure the user roles in your app to match the roles returned by SAM. End-users will be given the matching role when they sign into the app. If the role in the SAM token is not found in the Mendix app the end-user will be given the role `User`.
1. Save the configuration.

#### 8.2.2 Parsing PIB Access Tokens

Mendix has released a Private IAM Broker (PIB). This allows non-Mendix clients to request access tokens which authorize them to consume APIs exposed by your Mendix backend app without a dependency on any other IAM infrastructure.

> The OIDC SSO module provides a default access token parsing microflow for PIB. To use it, select the appropriate access token parsing microflow:
> 
> * For PIB, the default access token parsing microflow is `OIDC.Default_PIB_TokenProcessing_CustomATP`.
> * You can also parse the access token using the custom microflow `OIDC.ACT_Token_CustomATPRetrieveRoles` which takes an access token as input and returns a list of non-persistent objects of type `OIDC.Role`. See the next section for more information.

Author: ```suggestion
The OIDC SSO module provides a default access token parsing microflow for PIB. To use it, select the `OIDC.Default_PIB_TokenProcessing_CustomATP` microflow.
```

If you want to use the information in an access token which is a JWT, you need to parse the access token in a microflow. For example, you may want to assign user roles in your app based on the contents of the access token JWT.

* The OIDC module provides you with default microflows for parsing access tokens from the following IdPs:
    * Siemens SAM – in this case the `sws.samauth.role.name` claim is interpreted — for example:

        ```json {linenos=false}
        "sws.samauth.role.name": [
        "c1c31b36-2779-4ddd-a6e7-eaff22ad382c"
        ]
        ```

    * Microsoft Azure AD – in this case the `roles` claim is interpreted, using the roles claim in the access token — for example:

        ```json {linenos=false}
        "roles": [
        "c1c31b36-2779-4ddd-a6e7-eaff22ad382c"
        ]
        ```

    * Private IAM Broker (PIB) – in this case the `scope` claim is interpreted — for example:

        ```json {linenos=false}
        {
        "scope": "mx:app:userrole:a2275396-a31a-4f06-81c5-10fdfcb9f507 mx:app:userrole:c8de23c9-64fa-4faa-a263-6debc7e01edf mx:app:userrole:20c79248-e9f2-4131-b4a1-ee6b634b8f69"
        }
        ```

> 

Reviewer: I think it's usefull to have these fragments :-D

5. Configure the user roles in your app to match the roles returned by SAM. End-users will be given the matching role when they sign into the app. If the role in the SAM token is not found in the Mendix app the end-user will be given the role `User`.
6. Save the configuration.

#### 8.3.2 Parsing PIB Access Tokens

The OIDC SSO module provides a default access token parsing microflow for PIB. To use it, do the following.

1. Create a secure REST API endpoint following the instructions in [API Authentication](#api-authentication), above.
1. Run your app and sign in as an administrator.
1. Configure the PIB client information in the OIDC Client configuration screen.
1. Check **Enable Access Token Parsing** to parse access tokens when performing OIDC Client Configuration

> 1. Select the appropriate access token parsing microflow:

Reviewer: I think these 5 steps apply whichever Access Token parsing microflow is chosen. I think it should go up to the introduction part of section 8.3 (fka 7.3).

In this case the word "PIB" can be left out.

Author: Done: https://github.com/mendix/docs/pull/5449/commits/69d100a4b2f0a3e79168fe35af4ef5aae5ecbf32

To confirm that the authorization is working, get an access token from PIB and pass it to the API Endpoint using the authorization header. You can use Postman or any client application.

#### 8.2.3 Parsing Microsoft Azure AD Access Tokens

The OIDC SSO module provides a default access token parsing microflow for Azure AD. To use it, select the appropriate access token parsing microflow:

* For Azure AD, the default access token parsing microflow is `OIDC.Default_Azure_TokenProcessing_CustomATP`.

> To confirm that the authorization is working, get an access token from your Azure AD IdP and pass it to the API Endpoint using the authorization header. You can use Postman or any client application.

Reviewer: I would suggest leaving out the remark "You can ...application".
Or move that to the beginning of the section on access token parsing. It is guidance to the lowcode developer and it applies to each of the IdP's that may be used to issue the tokens.

Author: Moved it to the Testing and Troubleshooting section.
